### PR TITLE
feat: Automatic publication of admins

### PR DESCRIPTION
**Pull request by redwan-faris**
This is a:

- [x] ✨ **New feature** - new behaviour has been implemented
- [ ] 🐛 **Bug fix** - existing behaviour has been made to behave
- [ ] ♻️ **Refactor** - the behaviour has not changed, just the implementation
- [ ] ✅ **Test backfill** - tests for existing behaviour were added but the behaviour itself hasn't changed
- [ ] ⚙️ **Chore** - maintenance task, behaviour and implementation haven't changed

### Description

- **Purpose** -  the admin can add the suggestion  and its will be published  automatically without going to draft

- **How to check** - 
login as admin and then add the suggestion and check it in the publication directly
### Author checklist

- [x] I have written a title that reflects the relevant ticket
- [x] I have written a description that says what the PR does and how to validate it
- [ ] I have linked to the project board ticket (and any related PRs/issues) in the Links section
- [ ] I have added a link to this PR to the ticket
- [x] I have made the PR to `main` from a branch named `<category>/<name>`, e.g. `feature/edit-spaceships` or `bugfix/restore-oxygen`
- [x] I have manually tested that the app still works correctly
- [x] I have requested reviewers here and in my team chat channel

- [x] I have spoken with my PM or TL about any parts of this task that may have become out-of-scope, or any additional improvements that I now realise may benefit my project
- [x] I have added tests, or new tests were not required
- [x] I have updated any documentation (e.g. diagrams, schemas), or documentation updates were not required

### CODE REVIEW

**Comment by textbook** [comment]
Please do review the PR template in the description and fill it out accordingly.

**Review by textbook — CHANGES_REQUESTED** [review]
The basic functionality (when an admin posts a resource, publish it immediately) works OK. But looking at _how_ that's done, it's clear the suggested flow (`.github/CONTRIBUTING.md`) wasn't followed. The checklist in the PR template is intended as a reminder to actually _do_ those things - link the issue, write the tests, update the docs, ...

## Functionality

✅ When I suggest a resource as a non-admin, it is created as a draft. When I do it as an admin, it's immediately published.

💡 UI is misleading when an admin is using the Suggest form - it still tells them their submission will be reviewed by an admin:

Screenshot 2024-06-13 at 10 47 25

## Implementation

❓ Are we doing too much in the controller - maybe the _service_ should be deciding whether to create a draft or published resource (i.e. receiving both information about the resource _and the person creating it_, something like `service.create(resource, user)`) as that's part of our business logic not specific to to the transport layer.

❌ No E2E test

❌ No API integration test

❓ Do the docs need updating to reflect the changed behaviour of the endpoint?

### Key

- ✅ Nothing to change, just commenting
- 💡 Possible change, or suggestion to think about in the future
- ❓ Discussion required before approval
- ❌ Change required before approval

**Comment by redwan-faris** [comment]
> The basic functionality (when an admin posts a resource, publish it immediately) works OK. But looking at _how_ that's done, it's clear the suggested flow (`.github/CONTRIBUTING.md`) wasn't followed. The checklist in the PR template is intended as a reminder to actually _do_ those things - link the issue, write the tests, update the docs, ...
> 
> ## Functionality
> ✅ When I suggest a resource as a non-admin, it is created as a draft. When I do it as an admin, it's immediately published.
> 
> 💡 UI is misleading when an admin is using the Suggest form - it still tells them their submission will be reviewed by an admin:
> 
> Screenshot 2024-06-13 at 10 47 25
> 
> ## Implementation
> ❓ Are we doing too much in the controller - maybe the _service_ should be deciding whether to create a draft or published resource (i.e. receiving both information about the resource _and the person creating it_, something like `service.create(resource, user)`) as that's part of our business logic not specific to to the transport layer.
> 
> ❌ No E2E test
> 
> ❌ No API integration test
> 
> ❓ Do the docs need updating to reflect the changed behaviour of the endpoint?
> 
> ### Key
> * ✅ Nothing to change, just commenting
> * 💡 Possible change, or suggestion to think about in the future
> * ❓ Discussion required before approval
> * ❌ Change required before approval


Thank you for your time and effort in reviewing my code. I have refactored the code in the API section to make it clearer, placing each piece of code in its appropriate file. Additionally, I have modified the responses that appear on the client side.

However, as a backend developer, I am facing problems with client-side testing.

Could you help me with changing the authentication for the user? The test is failing because it signs in using the admin account by default, but I want to control the sign-in process so I can create tests for both types of roles in the system.

Thanks.

**Comment by textbook** [comment]
> Could you help me with changing the authentication for the user? The test is failing because it signs in using the admin account by default, but I want to control the sign-in process so I can create tests for both types of roles in the system.

What test is failing, and what exactly is it saying? You can push work in progress to the branch, then it'll run in the pipeline and we can all see it.

Across the various existing E2E cases there are examples of signing in using both admin _and non-admin_ users, so testing both roles should be resonably straightforward.

**Comment by redwan-faris** [comment]
> > Could you help me with changing the authentication for the user? The test is failing because it signs in using the admin account by default, but I want to control the sign-in process so I can create tests for both types of roles in the system.
> 
> What test is failing, and what exactly is it saying? You can push work in progress to the branch, then it'll run in the pipeline and we can all see it.
> 
> Across the various existing E2E cases there are examples of signing in using both admin _and non-admin_ users, so testing both roles should be resonably straightforward.

I apologize for the delayed response. This week, I am celebrating Eid, an important holiday in Iraq , 
e2e is ok for me and the testing is working now 
I am talking about the testing in the client side 
how ever I will push the code as soon as possible 
Thank you

**Comment by redwan-faris** [comment]
> > > Could you help me with changing the authentication for the user? The test is failing because it signs in using the admin account by default, but I want to control the sign-in process so I can create tests for both types of roles in the system.
> > 
> > 
> > What test is failing, and what exactly is it saying? You can push work in progress to the branch, then it'll run in the pipeline and we can all see it.
> > Across the various existing E2E cases there are examples of signing in using both admin _and non-admin_ users, so testing both roles should be resonably straightforward.
> 
> I apologize for the delayed response. This week, I am celebrating Eid, an important holiday in Iraq , e2e is ok for me and the testing is working now I am talking about the testing in the client side how ever I will push the code as soon as possible Thank you

@textbook

**Comment by textbook** [comment]
I have seen those messages. Do you have a specific question? You can see the tests failing [in the pipeline](https://github.com/CodeYourFuture/tech-products-demo/actions/runs/9574856619/job/26455735257), if that's what you're asking about - they give a lot of feedback, and I don't think _"failing because it signs in using the admin account by default"_ is a correct interpretation of what's happening - the client-side tests don't sign in at all, the network layer is mocked with MSW.

File: client/src/pages/Suggest/Suggest.test.jsx
Comment (textbook) [inline client/src/pages/Suggest/Suggest.test.jsx:37]: ❌ You should pretty much never have conditional logic in tests. Only _one_ of these cases is ever actually tested, there's no test where _"suggesting"_ would be shown (which is what made them start failing in the first place), so including the other branch just confuses the issue and implies that's tested when it isn't.

❌ It's also incorrect - whether it shows "suggesting" or "publishing" isn't a matter of the _request_ body (which never has a draft property - this should be clear from the assertions on it), it's a matter of the _response_ body (which is an empty object in all of these tests).

File: client/src/pages/Suggest/index.jsx
Comment (textbook) [inline client/src/pages/Suggest/index.jsx:51]: 💡 Typos in UI text
```suggestion
					? "Please use the form below to submit a resource. It will appear immediately on the home page because you are administrator."
```

File: e2e/integration/suggest.test.js
Comment (textbook) [inline e2e/integration/suggest.test.js:138]: ❌ The `PATCH` request is to _publish_ the resource, so that it shows up on the home page, which doesn't make sense in a test where: 1. the resource is already published (because the user is an admin); and 2. the home page isn't subsequently visited anyway.

File: server/resources/resources.test.js
Comment (textbook) [inline server/resources/resources.test.js:389]: 💡 This should really appear under the existing `POST /` group of tests; you can _nest_ `describe` blocks where you really need separate contexts:

```javascript
describe("/api/resources", () => {
  describe("POST /", () => {
    describe("as a user", () => {
      // ...
    });
  
    describe("as an admin", () => {
      // ...
    });
  });
});
```

File: e2e/integration/suggest.test.js
Comment (textbook) [inline e2e/integration/suggest.test.js:117]: 💡 It seems odd to visit the page, _then_ update the DB, then login, then _reload_ the home page (which you're on after logging in), maybe:
```suggestion
	cy.task("seed", { users: [admin] });
	cy.visit("/");
	cy.logInAs("admin@codeyourfuture.io");
```

File: client/src/components/Header/index.jsx
Comment (textbook) [inline client/src/components/Header/index.jsx:27]: ❓ This is interesting, not sure how I feel about this. You're right that the admin is no longer _suggesting_ as such, but the route is still `/suggest`. At this point maybe publish and suggest should be treated as separate routes and pages entirely, that just share the core form?

File: client/src/pages/Suggest/index.jsx
Comment (textbook) [inline client/src/pages/Suggest/index.jsx:17]: 💡 or maybe
```suggestion
	const isAdmin = user?.is_admin;
```

File: client/src/pages/Suggest/index.jsx
Comment (textbook) [inline client/src/pages/Suggest/index.jsx:33]: ✅ This is neat, basing it on the response means it continues to work even if the logic around roles changes later on.

File: client/src/pages/Suggest/index.jsx
Comment (textbook) [inline client/src/pages/Suggest/index.jsx:48]: 💡 or
```suggestion
			<h2>{isAdmin ? "Publish" : "Suggest"} a resource</h2>
```

File: client/src/pages/Suggest/index.jsx
Comment (textbook) [inline client/src/pages/Suggest/index.jsx:1]: 💡 Continuing from the comment on the nav, this now has lots of separate checks of the same thing, which is a bit awkward. It'd be neater to have one higher-level e.g. `isAdmin ? <Publish {...} /> : <Suggest {...} />`.

File: server/resources/resources.test.js
Comment (textbook) [inline server/resources/resources.test.js:414]: 💡 Most of this is already asserted on in other tests, I wonder if the core logic being tested here would be clearer if the assertion was more focused:

```suggestion
			expect(body).toHaveProperty("draft", false);
```

**Review by textbook — COMMENTED** [review]
Thank you for the iteration, the test coverage is much improved. How did you find writing them? What about the refactoring from the controller to the service, what did you think of that?

> is there any project linked with this repo to link the ticket?

You can just link the issue, #11

**Comment by textbook** [comment]
This is now complete from the application perspective.

File: client/src/pages/Suggest/Suggest.test.jsx
Comment (redwan-faris) [reply client/src/pages/Suggest/Suggest.test.jsx:37]: ok I got that thanks

File: e2e/integration/suggest.test.js
Comment (redwan-faris) [reply e2e/integration/suggest.test.js:138]: ohhh that right :( mybe I forgot to delete it

File: client/src/components/Header/index.jsx
Comment (redwan-faris) [reply client/src/components/Header/index.jsx:27]: ya I thought about that and about the component name 
But I did'nt want to make major updates on the code without reviewing the current

**Comment by redwan-faris** [comment]
thank you for your effors and comments that I will sure work on them in the future

**Comment by redwan-faris** [comment]
> Thank you for the iteration, the test coverage is much improved. How did you find writing them? What about the refactoring from the controller to the service, what did you think of that?
> 
> > is there any project linked with this repo to link the ticket?
> 
> You can just link the issue, #11

I have reviewed the client testing and end-to-end (e2e) testing, and I found them to be very useful for completing the task. Regarding the controller logic, I realized I had misunderstood the project's code structure. I initially thought all logic should be placed in the service layer, but I noticed some validation and checks being performed in the controller as well. However, I now understand that the correct approach is indeed to place most of the business logic in the service layer.

**Pull request closed without merging**
